Prepare IP docs for incubation (#38302)
gradle/gradle: platforms/documentation/docs/src/docs/userguide/optimizing-builds/isolated_projects.adoc

When the Isolated Projects feature is enabled, the projects in a <<multi_project_builds.adoc#multi_project_builds,multi-project build>> are "isolated" from each other.
This means that build logic, such as build scripts or plugins, applied to a project cannot directly access the mutable state of other projects.
This constraint allows safely running project configuration in parallel, improving performance.

> Further optimizations, such as finer-grained caching of configuration results, are possible but are not yet available.

Author: 💭 Not sure whether it's a good idea to mention the non-implemented feature here. But it's also good to highlight that there is more coming, without committing to it.

> Further optimizations, such as finer-grained caching of configuration results, are possible but are not yet available.

Reviewer: ```suggestion
Further optimizations, such as finer-grained caching of configuration results, are possible but are not yet implemented.
```

* Tooling model caching is local only: no sharing across checkouts or remote caching.

[[sec:evaluating]]
== Evaluating Isolated Projects for your build

The benefit Isolated Projects brings depends on the size and shape of your build, your hardware, and your workflow (see <<sec:performance,performance benefits and limitations>>).
Rather than predicting the gain from the structure of your build, the most reliable approach is to measure it directly:

. Enable the feature
+
Make the build <<configuration_cache_enabling.adoc#config_cache:adoption,compatible with Configuration Cache>> first, then enable Isolated Projects.

. Get an early estimate before completing the migration
+
You don't have to fix every <<sec:constraint_violations,violation>> before measuring.
Use the <<sec:dangerously_ignore_problems,_dangerously ignore problems_ mode>> to gauge the speed-up for your actual workflows — both running tasks and IDE sync — while violations are still present.

. Compare configuration-phase times
+
Measure the configuration phase with and without the feature using a <<inspect.adoc#inspecting_build_scans,Build Scan or profile report>>.

> Keep the ceiling in mind: Isolated Projects can only speed up the <<build_lifecycle.adoc#configuration,configuration phase>>, so its upper bound is the share of build time that phase represents.

Reviewer: ```suggestion
Keep in mind that Isolated Projects can only speed up the <<build_lifecycle.adoc#configuration,configuration phase>>, so it can only improve the build time that phase represents. The <<build_lifecycle.adoc#execution,execution phase>> duration is unaffected.
```

[[sec:project_to_build_access]]
=== Project-to-build access constraints

Project configuration logic has access to the build-scoped state shared across all projects via the
link:{javadocPath}/org/gradle/api/invocation/Gradle.html[`Gradle`] object, reachable through `project.gradle`.
Each `Gradle` instance exposes both immutable build information and mutable build-scoped state.
Accessing the mutable state from a project results in a violation, while accessing immutable build information is allowed.

A large part of the `Gradle` interface is dedicated to callback registration.
Registering these callbacks from a project couples that project to build-scope state and to the lifecycle of other projects, which is not safe when projects are configured in parallel.
Callbacks that do get successfully registered may also execute in a different order per invocation.

[[sec:mutable_project_to_build]]
==== Restricted access to mutable build state

> Registering listeners or callbacks, applying plugins, or mutating extensions on the `Gradle` object from a project is not allowed with Isolated Projects, and produces a violation such as:

Reviewer: ```suggestion
Registering listeners or callbacks, applying plugins, or mutating extensions on the `Gradle` object from a project is not allowed with Isolated Projects, and produces a violation like:
```

* Use the latest tooling, such as the latest versions of Android Studio and IntelliJ IDEA
+
Newer IDE versions can bring better performance as well as a better user experience when adopting Isolated Projects.

> A recommended version of IntelliJ IDEA is 2026.1.1 or later.

Reviewer: ```suggestion
It is recommended to use IntelliJ IDEA 2026.1.1 or later.
```

* Project configuration runs *sequentially* instead of in parallel.
* Configuration Cache store runs *sequentially*.
* Per-project model reuse across builds is *disabled*, so every model is rebuilt after an input change.

> * All projects are configured, ignoring any Configuration on Demand setting.

Reviewer: ❓ Should we link to the relevant "Configuration on Demand" docs here?

The Configure on Demand flag is _ignored_ when Isolated Projects is enabled.

[[sec:faq]]
== Frequently asked questions

> Common questions about adopting Isolated Projects.

Reviewer: I think we might be able to leave this out?

However, the current implementation doesn't leverage that yet.
Instead, all projects are configured in parallel unless the entire configuration phase is skipped on a Configuration Cache hit.

> The Configuration on Demand flag is _ignored_ when Isolated Projects is enabled.

Reviewer: ❓ Also link to "Configuration on Demand" here?

[[sec:faq]]
== Frequently asked questions

[[sec:faq_project_dependencies]]
=== Does a complex project dependency graph limit the benefit of Isolated Projects?

No.
A common misconception is that Isolated Projects optimizes builds by following dependencies between projects, so that a deep or highly-connected dependency graph would limit the benefit.

The benefit comes from <<sec:performance,parallelizing the configuration phase>>.

> It is driven by the number of projects and the cost of configuring them.

Reviewer: 🤔 Maybe we should address the specific concern here more directly, by adding a sentence like this?
```suggestion
The benefit comes from <<sec:performance,parallelizing the configuration phase>>.
This phase does not require resolving dependencies between projects, so it is not limited by them.
The performance is only affected by the number of projects and the cost of configuring them.
```
